Add new compressor methods to Run.getLogStream()

HedAurabesh opened the pull request:
Currently, Run.getLogStream() only supports uncompressed and GZIP compressed
log files.

This change adds support for more compression methods from Apache Commons:
 - XZ
 - LZMA
 - Bzip2
 - Snappy

This allows the Jenkins users to choose from a wider range of compression
algorithms for their logfiles -- tuning Jenkins either for better CPU usage or
for smaller log files.

Signed-off-by: Martin Schroeder <martin.h.schroeder@intel.com>

HedAurabesh committed:
Add new compressor methods to Run.getLogStream()

Currently, Run.getLogStream() only supports uncompressed and GZIP compressed
log files.

This change adds support for more compression methods from Apache Commons:
 - XZ
 - LZMA
 - Bzip2
 - Snappy

This allows the Jenkins users to choose from a wider range of compression
algorithms for their logfiles -- tuning Jenkins either for better CPU usage or
for smaller log files.

Signed-off-by: Martin Schroeder <martin.h.schroeder@intel.com>

oleg-nenashev commented on core/src/main/java/hudson/model/Run.java:
```diff
@@ -1,6 +1,8 @@
  * The MIT License
- * 
+ *
+ * Copyright (c) 2019 Intel Corporation 
```
It is better to modify the copyright below and to just add Intel to the list

HedAurabesh replied:
Okay, I will adjust it appropriately.
This code was lifted from internal code changes, where the policy is that the newest copyright has to go on top. :)

As for the implementation as an extension:  We considered the same, and created a larger change, which also enables Jenkins to compress the logs on its own at the end of the build.

If you prefer, we can upload that larger change, instead of this smaller change.
Our assumption was just that this smaller change might be easier to get upstreamed first, before we come in with the larger alteration.

oleg-nenashev reviewed (approved):
Current implementation should not cause much performance overheads, but it might be better to extract extension and to fetch from the map instead

HedAurabesh committed:
Changed copyright message as per Oleg's recommendation.

n/t

batmat reviewed (commented):
LGTM

daniel-beck commented:
Isn't this API essentially deprecated due to JEP-210?

Besides that, I would prefer a well-reasoned selection of algorithms rather than whatever formats are currently available in the commons-compress library (which is the impression I get here). If the library were to drop support for one of them, or we need to switch to a different library, we may have problems retaining support for the formats specified here.

daniel-beck reviewed (changes requested):
See previous comment.

jvz reviewed (approved):
I like the idea. Some points:

* Would be great for some documentation updates here.
* I'm not so sure about bzip2 support. That format is more useful for compressing data at rest, and even that is basically superseded by formats like xz.

daniel-beck commented:
@HedAurabesh Thanks for your response!

> Compression : XZ/LZMA > GZIP > Snappy > Uncompressed

For use here, I would expect the decompression speed to be more important? Since compression would necessarily be out of band anyway? Or is the order the same?

HedAurabesh commented:
> > Compression : XZ/LZMA > GZIP > Snappy > Uncompressed
> 
> For use here, I would expect the decompression speed to be more important? Since compression would necessarily be out of band anyway? Or is the order the same?

The order for decompression changes, depending on how much entropy/redundancy the input file has.

For example, here are the numbers for handling:
 - 200 MB of highly repetitive log text (99% redundancy)
 - 200 MB of Wikipedia HTML (~80% redundancy)
 - 128 MB of random data (0% redundancy)

All values are in MB/s and the compression level is set to the default (usually L=6).

```
Algo   |      Log      |      Wiki     |     Random    |
       |   C   |   D   |   C   |   D   |   C   |   D   |
-------+-------+-------+-------+-------+-------+-------|
Snappy | 697.0 | 626.0 |  52.9 | 149.0 | 657.0 |1792.0 |
GZIP   | 131.0 | 320.0 |  17.2 | 113.0 |  24.4 | 175.0 |
XZ     |  16.9 | 569.0 |   1.5 |  52.0 |   2.6 | 438.0 |
LZMA   |  17.2 |1638.0 |   1.7 |  60.7 |   2.2 |  12.9 |
BZIP2  |   2.2 |  67.0 |   9.8 |  24.9 |   5.6 |  13.1 |
```

As you can see, the different compressors use different optimizations for the extreme cases of redundancy. But, as far as compression is concerned, there is mostly a stable ratio (unless you mess with the compressor levels).

For compression, faster compressors are faster, irrespective of redundancy.
The exception is BZIP2 which shows unpredictable speed because of its use of the Burrows-Wheeler transform.

But in terms of decompression speed, the type of input makes a huge difference, which makes it difficult to give a nicely sorted overview. :D


That said, I would expect most log files to behave more like the Wikipedia, and less like the two extremes.

In that case, a nice decompression order can be given and it is the same as for compression, where "<" means "is slower than":
 - (BZIP2 <) XZ/LZMA < GZIP < Snappy < Uncompressed


As for the compression size (values in Bytes):
```
Algo   |    Log            |    Wiki           |  Random   |
-------+-------------------+-------------------+-----------|
None   | 203226708         | 209715200         | 134217728 |
Snappy |  10824332 (94.7%) | 112479826 (46.4%) | 134234122 |
GZIP   |   1338809 (99.3%) |  72067579 (65.6%) | 134238893 |
XZ     |     67312 (99.9%) |  52865856 (74.8%) | 134224436 |
LZMA   |     66364 (99.9%) |  52857630 (74.8%) | 136039645 |
BZIP2  |    268259 (99.9%) |  58175931 (72.3%) | 134810169 |
```

HedAurabesh committed:
Removed BZIP2 support and added distinction between the two snappy formats

Following the recommendations, this commit drops the BZIP2 format, since
it seems not sensible for log files, given the newer compression formats.

Also, this commit makes the distinction between the framed and raw snappy
formats, that are usually distinguished by their extension.

The framed format also has a magic number prefix, but for that the first few
bytes of the file or stream would need to be read.

HedAurabesh commented:
> jvz left a comment:
> * Would be great for some documentation updates here.

Which documentation would need to be updated, ideally?
The Jenkins Wiki? The inline help?

I presume just writing this in the changelog would vanish too quickly.

jvz commented:
I'm not actually sure where this would best go. Perhaps as an inline help file?

daniel-beck commented:
> Perhaps as an inline help file?

There's no UI for this anywhere (unless you could https://plugins.jenkins.io/compress-buildlog which you shouldn't).

Beyond that, I don't think this is documented anywhere "official" to begin with. Asking @HedAurabesh to start the non-existing doc is a bit much (but would of course be appreciated).

HedAurabesh committed:
Removed dependency on deprecated getLogFile() method

The previous code was missing a change to getLogFile(), that was
not applied due to the deprecation of getLogFile() in the upstream code.

This led to the new files never being read. The new code will pick up
compressed log files regardless of their extension, as long as they are
named "log" or "log.*" -- with the uncompressed version always being
preferred.

Also:
 - Extended the logging to give some additional feedback to the users
   in case of errors or missing files
 - Rephrased the API docs to no longer refer to log files
   (See: https://github.com/jenkinsci/jenkins/pull/3724)

HedAurabesh commented:
I have uploaded a new commit, that fixes an issue due to the deprecation of getLogFile(). The previous submission was accidentally incomplete due to that difference between our internal code and the upstream code.

@jvz , @daniel-beck : I also extended the logging and feedback to the user in case of issues. That's still not proper documentation, of course.

jglick commented:
I do not think we should do this. See [JEP-207](https://jenkins.io/jep/207) for non-Pipeline builds.

HedAurabesh committed:
Fix error reported by findbugs.

n/t

HedAurabesh commented:
> I do not think we should do this. See [JEP-207](https://jenkins.io/jep/207) for non-Pipeline builds.

Hi Jesse. Given that JEP-207 looks like a major rewrite of the entire logging system, I am not quite sure why this change would collide with that?

Given that even post-JEP-207 the system will need to provide support for on-master log files, it would be no difference whether you migrate the old code or this new code here to JEP-207.

The difference between the two is in making use of a library that is already greatly in use with Jenkins anyway (Apache Commons) and the addition of handful of lines to handle files other than "log" and "log.gz".


Also, I see that JEP-207 is currently in "deferred" status, which might mean it can be months or years before it is worked on -- yet alone released to the public.

oleg-nenashev reviewed (approved):
As a former JEP-207 sponsor, I cannot say whether JEP-207 is going to be revived anytime soon. https://github.com/jenkinsci/jenkins/pull/3575 will definitely need a merge conflict fix if t is reopened, but it is not a big deal. So weak +1 from me for merging if there are multiple Jenkins users who demand it

HedAurabesh commented:
Looks like a new build of my change was started and fails for JDK11.

Is there anything that I need to adjust in this change, or is it (like I presume) that that the JDK11 checks are simply expected to fail at the moment?

daniel-beck commented:
That's environmental:

> Failed to install https://download.java.net/java/GA/jdk11/9/GPL/openjdk-11.0.2_linux-x64_bin.tar.gz to /home/jenkins/tools/hudson.model.JDK/jdk11

jglick commented:
> I am not quite sure why this change would collide with that?

JEP-207 is the right way to make customizations like this one—in, for example, the existing `compress-buildlog` plugin. I just do not think we should be adding more complexity and more untested operational modes to Jenkins core for this purpose.

oleg-nenashev commented:
@daniel-beck your feedback about the requested change responses would be appreciated

HedAurabesh commented:
I have taken a closer look at the ancillary code that interacts with the log in Run.java and there are plenty of them that assume an uncompressed log file and which only by voodoo magic works with gzip compressed log files. They fail in colourful ways when fed with other compression formats.

As such, I will close this pull request and instead upload our much more capable extension-point based interface in a while.

Thanks for your feedbacks on this PR.
